File: client/host/src/main/java/com/reservation/host/auth/controller/AuthController.java
package com.reservation.host.auth.controller;

import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import com.reservation.auth.annotation.LoginUserId;
import com.reservation.auth.login.LogoutService;
import com.reservation.auth.login.RefreshService;
import com.reservation.auth.login.Role;
import com.reservation.auth.login.dto.AccessTokenHeader;
import com.reservation.auth.login.dto.LoginSettingToken;
import com.reservation.auth.login.dto.RefreshTokenCookie;
import com.reservation.domain.host.Host;
import com.reservation.host.auth.controller.request.LoginRequest;
import com.reservation.host.auth.service.AuthService;
import com.reservation.support.response.ApiResponse;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@Tag(name = "인증 API", description = "숙박 업체 인증 API 입니다.")
@RequiredArgsConstructor
public class AuthController {
	public static final String PRE_AUTH_ROLE_HOST = "hasRole('ROLE_HOST')";

	private final AuthService authService;
	private final RefreshService refreshService;
	private final LogoutService logoutService;

	@PostMapping("/no-auth/login") //❗JWT auth 제외
	@Operation(summary = "일반 로그인", description = "숙박 업체 로그인 API 입니다, 로그인 성공 시 JWT 토큰을 발급합니다")
	public ResponseEntity<Void> login(@Valid @RequestBody LoginRequest request) {
		LoginSettingToken loginSettingToken = authService.login(request.email(), request.password());

		RefreshTokenCookie refreshTokenCookie = loginSettingToken.refreshTokenCookie();

		ResponseCookie responseCookie = ResponseCookie.from(refreshTokenCookie.name(), refreshTokenCookie.value())
			.httpOnly(true)
			.secure(true)
			.path("/")
			.maxAge(refreshTokenCookie.refreshDuration())
			.build();

		AccessTokenHeader accessTokenHeader = loginSettingToken.accessTokenHeader();

		return ResponseEntity.noContent()
			.header(accessTokenHeader.name(), accessTokenHeader.value())
			.header(HttpHeaders.SET_COOKIE, responseCookie.toString())
			.build();
	}

	@GetMapping("/no-auth/refresh") //❗JWT auth 제외 -> @LoginUserId 토큰 검증
	@Operation(summary = "토큰 재발급", description = "숙박 업체 AT 재발급 API 입니다, 기존 토큰은 만료됩니다")
	public ResponseEntity<Void> tokenReissue(@LoginUserId(includeExpired = true) long hostId) {
		AccessTokenHeader accessTokenHeader = refreshService.tokenReissue(hostId, Role.HOST);

		return ResponseEntity.noContent()
			.header(accessTokenHeader.name(), accessTokenHeader.value())
			.build();
	}

	@PostMapping("/auth/logout")
	@PreAuthorize(PRE_AUTH_ROLE_HOST)//✅숙박 업체만 접근 가능
	@Operation(summary = "로그아웃", description = "숙박 업체 로그아웃 API 입니다")
	public ResponseEntity<Void> logout(@LoginUserId long hostId) {
		String deleteRefreshTokenCookieKey = logoutService.logout(hostId, Role.HOST);

		ResponseCookie deleteCookie = ResponseCookie.from(deleteRefreshTokenCookieKey, "")
			.httpOnly(true)
			.secure(true)
			.path("/")
			.maxAge(0)
			.build();

		return ResponseEntity.noContent()
			.header(HttpHeaders.SET_COOKIE, deleteCookie.toString())
			.build();
	}

	@GetMapping("/auth/me")
	@PreAuthorize(PRE_AUTH_ROLE_HOST)//✅숙박 업체만 접근 가능
	@Operation(summary = "ME", description = "숙박 업체 정보 확인 API 입니다")
	public ApiResponse<Host> getMe(@LoginUserId long hostId) {
		return ApiResponse.ok(authService.findMe(hostId));
	}
}

File: identity-service/src/main/kotlin/com/msa/identityservice/auth/controller/AuthController.kt
package com.msa.identityservice.auth.controller

import com.msa.identityservice.annotation.LoginUser
import com.msa.identityservice.auth.consts.AuthConstants.AUTH_HEADER_NAME
import com.msa.identityservice.auth.consts.AuthConstants.REFRESH_COOKIE_NAME
import com.msa.identityservice.auth.consts.AuthConstants.getAccessTokenHeaderValue
import com.msa.identityservice.auth.controller.request.LoginRequest
import com.msa.identityservice.auth.controller.request.LogoutRequest
import com.msa.identityservice.auth.controller.response.LogoutAllResponse
import com.msa.identityservice.auth.controller.response.SessionInfoResponse
import com.msa.identityservice.auth.controller.response.SessionInfosResponse
import com.msa.identityservice.auth.service.AuthService
import com.msa.identityservice.auth.token.dto.LoginAuthToken
import com.msa.identityservice.auth.token.dto.TokenAuthInfo
import com.msa.identityservice.support.response.ApiResponse
import jakarta.validation.Valid
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseCookie
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*


@RestController
@RequestMapping("/auth")
class AuthController(
    private val authService: AuthService
) {

    @PostMapping("/login")
    fun login(
        @Valid @RequestBody
        request: LoginRequest
    ): ResponseEntity<Unit> {
        val loginRequestInfo = request.toLoginRequestInfo()
        val loginSettingToken = authService.login(loginRequestInfo)

        return setAuthTokenHeaders(loginSettingToken)
    }

    private fun setAuthTokenHeaders(loginAuthToken: LoginAuthToken): ResponseEntity<Unit> {
        val responseCookie: ResponseCookie = ResponseCookie.from(REFRESH_COOKIE_NAME, loginAuthToken.refreshToken)
            .httpOnly(true)
            .secure(true)
            .path("/")
            .maxAge(loginAuthToken.refreshTokenDuration)
            .build()

        return ResponseEntity.noContent()
            .header(AUTH_HEADER_NAME, getAccessTokenHeaderValue(loginAuthToken.accessToken))
            .header(HttpHeaders.SET_COOKIE, responseCookie.toString())
            .header("Access-Control-Expose-Headers", AUTH_HEADER_NAME)
            .build()
    }

    @GetMapping("/refresh")
    fun tokenReissue(): ResponseEntity<Unit> {
        val loginSettingToken = authService.tokenReissue()

        return setAuthTokenHeaders(loginSettingToken)
    }

    @GetMapping("/me")
    fun getMe(
        @LoginUser
        tokenAuthInfo: TokenAuthInfo
    ): ApiResponse<TokenAuthInfo> {

        return ApiResponse.read(
            message = "내 정보 조회에 성공했습니다.",
            data = tokenAuthInfo
        )
    }

    @GetMapping("/sessions")
    fun getSessions(): ApiResponse<SessionInfosResponse> {
        val sessionInfos = authService.getSessions()

        val sessions = sessionInfos.map { sessionInfo ->
            SessionInfoResponse(
                deviceId = sessionInfo.deviceId,
                loginDateTime = sessionInfo.loginDateTime,
                lastActivityDateTime = sessionInfo.lastActivityDateTime,
            )
        }

        return ApiResponse.read(
            message = "접속 중인 세션 조회에 성공했습니다.",
            data = SessionInfosResponse(sessions)
        )
    }

    @DeleteMapping("/logout")
    fun logout(): ResponseEntity<ApiResponse<SessionInfoResponse>> {
        val sessionInfo = authService.logout()

        val data = SessionInfoResponse(
            deviceId = sessionInfo.deviceId,
            loginDateTime = sessionInfo.loginDateTime,
            lastActivityDateTime = sessionInfo.lastActivityDateTime,
        )

        return deleteCookieResponse(
            message = "로그아웃 되었습니다.(${sessionInfo.deviceId})",
            data = data
        )
    }

    private fun <T> deleteCookieResponse(
        message: String,
        data: T
    ): ResponseEntity<ApiResponse<T>> {
        val deleteCookie = ResponseCookie.from(REFRESH_COOKIE_NAME, "")
            .httpOnly(true)
            .secure(true)
            .path("/")
            .maxAge(0)
            .build()

        val responseBody = ApiResponse.delete(
            message = message,
            data = data
        )

        return ResponseEntity.ok()
            .header(HttpHeaders.SET_COOKIE, deleteCookie.toString())
            .body(responseBody)
    }

    @DeleteMapping("/logout-device")
    fun logout(
        @Valid @RequestBody
        request: LogoutRequest
    ): ApiResponse<SessionInfoResponse> {
        val deviceId = request.deviceId!!
        val sessionInfo = authService.logout(deviceId)

        val data = SessionInfoResponse(
            deviceId = sessionInfo.deviceId,
            loginDateTime = sessionInfo.loginDateTime,
            lastActivityDateTime = sessionInfo.lastActivityDateTime,
        )

        return ApiResponse.delete(
            message = "로그아웃 되었습니다.($deviceId)",
            data = data
        )
    }

    @DeleteMapping("/logout-all")
    fun logoutAll(): ResponseEntity<ApiResponse<LogoutAllResponse>> {
        val sessionInfos = authService.logoutAll()
        val sessionInfoResponseList = sessionInfos.map { sessionInfo ->
            SessionInfoResponse(
                deviceId = sessionInfo.deviceId,
                loginDateTime = sessionInfo.loginDateTime,
                lastActivityDateTime = sessionInfo.lastActivityDateTime,
            )
        }

        return deleteCookieResponse(
            message = "접속 중인 모든 기기에서 로그아웃 되었습니다.",
            data = LogoutAllResponse(sessionInfoResponseList)
        )
    }

}
